### ISSUE TITLE
Pods dead because of Failed to update QoS cgroup

### ISSUE THREAD

**Issue by stibi** [kubernetes/kubernetes#49835]
/kind bug

After some time, usualy once a day, all the pods deployed on a cluster ends up a dead and kubelet logs a lot of error like this:

```
Jul 27 10:55:10 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:10.710794     855 kubelet.go:1912] SyncLoop (PLEG): "impressive-swan-nginx-ingress-controller-x1cm4_default(52bec2b1-7255-11e7-8313-06f61a7ca0ba)", event: &pleg.PodLifecycleEvent{ID:"52bec2b1-7255-11e7-8313-06f61a7ca0ba", Type:"ContainerDied", Data:"468fe549b5ec480796fafd6bdb7e80668a869fe20dd4f6256c8321b14dea1a75"}
Jul 27 10:55:10 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: W0727 10:55:10.711017     855 pod_container_deletor.go:77] Container "468fe549b5ec480796fafd6bdb7e80668a869fe20dd4f6256c8321b14dea1a75" not found in pod's containers
Jul 27 10:55:10 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:10.866742     855 kuberuntime_manager.go:500] pod "impressive-swan-nginx-ingress-controller-x1cm4_default(52bec2b1-7255-11e7-8313-06f61a7ca0ba)" container "nginx-ingress-controller" is unhealthy, it will be killed and re-created.
Jul 27 10:55:10 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:10.866814     855 kuberuntime_container.go:561] Killing container "docker://36001b967b8e23388173c1655488bd348c7f250b10aa62086adb0e23fc4a7a0e" with 60 second grace period
Jul 27 10:55:10 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:10.870280     855 kuberuntime_manager.go:741] checking backoff for container "nginx-ingress-controller" in pod "impressive-swan-nginx-ingress-controller-x1cm4_default(52bec2b1-7255-11e7-8313-06f61a7ca0ba)"
Jul 27 10:55:11 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:11.049386     855 qos_container_manager_linux.go:303] [ContainerManager]: Failed to update QoS cgroup configuration
Jul 27 10:55:11 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:11.049401     855 kubelet_pods.go:694] Failed to update QoS cgroups while killing pod: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 2 to cpu.shares: open /var/lib/rkt/pods/run/cb32f496-d289-412d-95a1-75fc8c2556bb/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
Jul 27 10:55:11 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:11.049578     855 qos_container_manager_linux.go:303] [ContainerManager]: Failed to update QoS cgroup configuration
Jul 27 10:55:11 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:11.049591     855 kubelet.go:1579] Failed to update QoS cgroups while syncing pod: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 2 to cpu.shares: open /var/lib/rkt/pods/run/cb32f496-d289-412d-95a1-75fc8c2556bb/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
Jul 27 10:55:11 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: I0727 10:55:11.579936     855 kube_docker_client.go:331] Stop pulling image "gcr.io/google_containers/nginx-ingress-controller:0.9.0-beta.11": "Status: Image is up to date for gcr.io/google_containers/nginx-ingress-controller:0.9.0-beta.11"
Jul 27 10:55:11 ip-10-20-3-11.eu-west-1.compute.internal kubelet[855]: E0727 10:55:11.628216     855 remote_runtime.go:208] StartContainer "81cae6e176a44459db48e47caf684025b481ae6c1531221ffaf3b886b9b0f475" from runtime service failed: rpc error: code = 2 desc = failed to start container "81cae6e176a44459db48e47caf684025b481ae6c1531221ffaf3b886b9b0f475": Error response from daemon: {"message":"cannot join network of a non running container: 468fe549b5ec480796fafd6bdb7e80668a869fe20dd4f6256c8321b14dea1a75"}
```

I'm not sure if it is related to flanneld, or what went wrong. I have to restart kubelet on all affected nodes and then everything works again for some time.

I have not experienced this problem on k8s 1.5.2, but I can see it on 1.6.7 and 1.7.2.

I don't have a reliable way how to reproduce this, I don't know what triggers the problem.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.2
- Cloud provider or hardware configuration**: AWS
- OS (e.g. from /etc/os-release): CoreOS 1409.2.0
- Kernel (e.g. `uname -a`): 4.11.9-coreos

**Comment by stibi** [comment]
@kubernetes/sig-node-bugs

**Comment by PengTaoWW** [comment]
@stibi look like cgroup change make this

**Comment by stibi** [comment]
yes, something about cgroups…but I can't find out what is wrong or how to reproduce it reliably. 

Maybe it's related only to flannel? (`failed to write 2 to cpu.shares: open /var/lib/rkt/pods/run/cb32f496-d289-412d-95a1-75fc8c2556bb/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory`)

Any hints how to approach this appreciated.

**Comment by stibi** [comment]
Maybe this is the same problem, I'll update once I find out more https://github.com/coreos/bugs/issues/1965

**Comment by resouer** [comment]
cc @derekwaynecarr Mind take a look?

**Comment by matt-duch** [comment]
Kubernetes version (use kubectl version): 1.7.2
Cloud provider or hardware configuration**: AWS
OS (e.g. from /etc/os-release): CoreOS/Container Linux 1409.7.0
Kernel (e.g. uname -a): 4.11.11-coreos

I'm running into this problem as well, except rather than being time based, I can trigger this behavior at will. I'm running with the fix mentioned in coreos/bugs#1965 (I've tried running without it as well), so I don't see the issue at certain times of day. I have a small staging cluster where I restrict certain jobs (using node selectors) to a few nodes. If I run enough jobs of that type, I'll trigger this error. What I've noted:
Jul 31 21:15:29 ip-10-0-2-88 kubelet[3671]: I0731 21:15:29.267611    3671 kubelet.go:1579] Failed to update QoS cgroups while syncing pod: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 2 to cpu.shares: open /var/lib/rkt/pods/exited-garbage/7ec267f8-5159-4297-90fb-3eac594a8480/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
1) This rkt pod is *not* managed by kubernetes, it's launched by a systemd service unit on all nodes, but the kubelet is still trying to set a value under a kubepods subpath. 
2) This container has been stopped (or removed, based on whether or not the mentioned fix has been applied), it no longer exists. 
3) This issue will persist indefinitely (as far as I can tell) until you restart the kubelet. It seems as if this pod is being cached in the kubelet and it's entering an error loop.

Basically, I think what's happening is: node resource pressure such as OOM errors (or something time based) is causing the flannel container to be stopped/removed (then restarted by systemd), but the kubelet is still trying to manage it. I think we might be seeing this with flannel because it's a non-kubernetes rkt pod.

**Comment by zxpower** [comment]
Kubernetes version (use kubectl version): 1.8.1
Cloud provider or hardware configuration**: Bare metal
OS (e.g. from /etc/os-release): CoreOS/Container Linux 1465.8.0
Kernel (e.g. uname -a): 4.12.14-coreos
Flannel: 0.9.0
Etcd: 3.0.17

We're running small 4 node (1 master. 3 workers) bare metal cluster that I'm preparing for production. Before this weekend nodes were crushing randomly, but on last Friday night they crashed all and went into infinite loop of killing of pods. Only after I restarted kubelet first on master and then on all nodes I got cluster back to normal operation.

On master node (crashed last from whole cluster):
```
Oct 21 01:21:31 core4 kubelet-wrapper[974]: W1021 01:21:31.071548     974 qos_container_manager_linux.go:145] [ContainerManager] Failed to reserve QoS requests: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 307 to cpu.shares: open /var/lib/rkt/pods/run/8b256cf8-7a52-4bce-9b12-e33d6eacc3b1/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
```

On one of worker nodes (this one failed first):
```
Oct 21 00:26:02 core5 kubelet-wrapper[969]: W1021 00:26:02.922022     969 qos_container_manager_linux.go:145] [ContainerManager] Failed to reserve QoS requests: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 2 to cpu.shares: open /var/lib/rkt/pods/run/638aee6a-9e5b-4325-b138-e649b960beef/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
```

Another node (crashed next):
```
Oct 21 01:08:50 core6 kubelet-wrapper[975]: W1021 01:08:50.477284     975 qos_container_manager_linux.go:145] [ContainerManager] Failed to reserve QoS requests: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 2 to cpu.shares: open /var/lib/rkt/pods/run/84d640dd-ec11-4c92-a21a-51f1ef03893c/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
```

Another node (crashed last from nodes):
```
Oct 21 01:13:43 core7 kubelet-wrapper[991]: W1021 01:13:43.402740     991 qos_container_manager_linux.go:145] [ContainerManager] Failed to reserve QoS requests: failed to set supported cgroup subsystems for cgroup /kubepods/burstable: Failed to set config for supported subsystems : failed to write 112 to cpu.shares: open /var/lib/rkt/pods/run/bf1878bb-5301-4249-aa75-3c25c582ecbe/stage1/rootfs/opt/stage2/flannel/rootfs/sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/cpu.shares: no such file or directory
```

**Comment by wladh** [comment]
I investigated this issue a bit and the problem is that `rkt` mounts cgroups inside its container rootfs (in `/var/lib/rkt/pods`) and they are visible in root container (you can do `cat /proc/self/mountpoints`). `cadvisor` (more specifically its `libcontainer.GetCgroupSubsystems`) picks them up and due to the ordering of entries in `mountfile` one of those mount points will become the root directory through which `cadvisor` interacts with cgroups on the system. That works fine until whatever runs in that `rkt` container dies, or the system reboots and `rkt gc` will move the container storage from `/var/lib/rkt/pods/run` to `/var/lib/rkt/pods/exited-garbage`. Then the paths to cgroups in `cadvisor` will be broken (and the same mechanism of getting cgroup mounts is used in other parts of `kubelet` as well).
I don't know enough about `rkt` to understand if that's something that can be fixed there or maybe `cadvisor/libcontainer.GetCgroupSubsystems` can be changed to filter out those mount points. I'll open an issue in `cadvisor`.

**Comment by euank** [comment]
Thanks for your investigation @wladh, I looked into cadvisor's code and it looks like it does reliably pick the last one, which is pretty much always the wrong one if there are multiple.

I've opened a PR at https://github.com/google/cadvisor/pull/1792 which might fix this.

**Comment by tuminoid** [comment]
We're hitting this still today with k8s 1.15.3, Docker 1.13.1 (+RHEL backports) on RHEL 7.7. Cleaning up kubepod slices do nothing, and so far only solution has been rebooting the node in question.
